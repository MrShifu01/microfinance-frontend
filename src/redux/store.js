import { configureStore } from '@reduxjs/toolkit';
import pageReducer from './pageSlice';
import clientsReducer from './clientsSlice';
import loansReducer from './loansSlice';
import userReducer from './userSlice';
import settingsReducer from './settingsSlice';

const store = configureStore({
  reducer: {
    settings: settingsReducer, // Reducer for managing settings state
    user: userReducer, // Reducer for managing user state
    loans: loansReducer, // Reducer for managing loans state
    clients: clientsReducer, // Reducer for managing clients state
    page: pageReducer, // Reducer for managing page state
  },
});

export default store;
