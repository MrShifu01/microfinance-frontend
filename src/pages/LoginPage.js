import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import axios from 'axios';
import { setUser, setTempUser } from '../redux/userSlice';
import { Link, Navigate } from 'react-router-dom'
import { Checkbox, FormControlLabel } from '@mui/material';
import { setPage } from "../redux/pageSlice";
import { setToken } from '../redux/tokenSlice';

export default function Login() {
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')
    const [redirect, setRedirect] = useState(false)
    const [rememberMe, setRememberMe] = useState(false); // State for the Checkbox

    const dispatch = useDispatch()

    useEffect(() => {
      dispatch(setPage('login')); // Set the page in the Redux store to 'login' when the component mounts
    }, [dispatch]);

    const handleLogin = async (ev, email, password) => {
        ev.preventDefault()
        try {
            const response = await axios.post('/users/login', {
              email,
              password
          })
              // Check if rememberMe is checked and dispatch accordingly
            if (rememberMe) {
              // Dispatch an action for rememberMe being checked
              dispatch(setUser(response.data));
            } else {
              // Dispatch an action for rememberMe being unchecked
              dispatch(setTempUser(response.data));
            }
            dispatch(setToken(response.data.token))
          setRedirect(true)
        } catch (error) {
          alert("Username or Password is incorrect, Try again.")
        }
    }

    if (redirect) {
      return <Navigate to='/'/>
    }

  return (
    <Box
      component="form"
      sx={{
        '& .MuiTextField-root': { m: 1, width: '25ch' },
      }}
      noValidate
      autoComplete="off"
      className='mx-auto mt-10'
    >
      <div className='flex flex-col'>
        <h2 className='text-2xl font-bold text-center mb-4'>Login</h2>
        <TextField
          required
          id="outlined-required"
          label="Email"
          value={email}
          onChange={e => setEmail(e.target.value)}
        />
        <TextField
          required
          id="outlined-disabled"
          label="Password"
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
        <FormControlLabel
          className='flex justify-center'
          label="Remember me?"
          labelPlacement="start"
          control={<Checkbox checked={rememberMe} onChange={ev => setRememberMe(ev.target.checked)} />}
        />
        <button className='bg-blue-500 py-1 px-3 w-20 mx-auto rounded-lg my-2' onClick={ev => handleLogin(ev, email, password)}>Login</button>
      </div>
      <p className='text-center text-sm text-gray-500'>Don't have an account? <Link to='/register' className='mx-2 underline text-black'>Register</Link></p>
      
      {/* For App testing Only */}
      <div className="my-6 border border-gray-500 p-4">
        <h2 className="text-sm font-semibold mb-2">Example Users:</h2>
        <ul className="list-disc ml-6">
          <li>
            <span className="font-semibold">Email:</span> johndoe@gmail.com
            <br />
            <span className="font-semibold">Password:</span> johndoe
            <br />
            <span className="font-semibold">Role:</span> Not Admin
          </li>
          <li className="mt-4">
            <span className="font-semibold">Email:</span> janesmith@gmail.com
            <br />
            <span className="font-semibold">Password:</span> janesmith
            <br />
            <span className="font-semibold">Role:</span> Admin
          </li>
        </ul>
      </div>

    </Box>
  );
}