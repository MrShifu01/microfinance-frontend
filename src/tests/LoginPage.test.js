import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import LoginPage from "../pages/LoginPage";
import { Provider } from "react-redux";
import store from "../redux/store";

jest.mock("axios", () => ({
  post: jest.fn(() => Promise.resolve({ data: {} })),
}));

describe("Jest Snapshot testing suite", () => {
  it("Matches DOM Snapshot", () => {
    const domTree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <LoginPage />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
