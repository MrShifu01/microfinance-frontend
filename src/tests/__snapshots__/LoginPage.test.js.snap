// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<form
  autoComplete="off"
  className="mx-auto mt-32 MuiBox-root css-13r4j3i"
  noValidate={true}
>
  <div
    className="flex flex-col"
  >
    <h2
      className="text-2xl font-bold text-center mb-4"
    >
      Login
    </h2>
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="outlined-required"
        id="outlined-required-label"
      >
        Email
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="outlined-required"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Email
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="outlined-disabled"
        id="outlined-disabled-label"
      >
        Password
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="outlined-disabled"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Password
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <label
      className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart flex justify-center css-1f2kvjf-MuiFormControlLabel-root"
    >
      <span
        className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          checked={false}
          className="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate={false}
          onChange={[Function]}
          required={false}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <span
        className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        Remember me?
      </span>
    </label>
    <button
      className="bg-primary w-20 mx-auto rounded-lg my-2"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
  <p
    className="text-center text-sm text-gray-500"
  >
    Don't have an account? 
    <a
      className="mx-2 underline text-black"
      href="/register"
      onClick={[Function]}
    >
      Register
    </a>
  </p>
</form>
`;
